### ISSUE TITLE
addCustomControlMapping for opening subMenu ?

### ISSUE THREAD

**Issue by vesper8** [php-school/cli-menu#171]
I can set keyboard mapping for regular callables but I would like to use it to open a submenu.. any way to achieve that?

```
        $subMenu = function (CliMenuBuilder $b) {
            $b->setTitle('Behold the awesomeness')
                ->addItem('Print Hello', function () { echo 'Hello!'; });
        };

        $exit = function (CliMenu $menu) {
            $menu->close();
        };

        $menu = (new CliMenuBuilder)
        ->addSubMenu('[S]uper Sub Menu', $subMenu)
        ->build();

        $menu->addCustomControlMapping('x', $exit);
        $menu->addCustomControlMapping('s', $subMenu);

        $menu->open();
```

**Comment by AydinHassan** [comment]
Unfortunately I don't think there is a way to do this. I think you could maybe hack it in somehow like this:

```php
$menu->addCustomControlMapping('s', function (CliMenu $menu) {
    foreach ($menu->getItems() as $item) {
        if ($item instanceof MenuMenuItem && $item-> getSubMenu()->getTitle() === 'Behold the awesomeness') {
            $item->showSubMenu($menu);
        }
    }
});
```

Not tested though!

**Comment by AydinHassan** [comment]
I'd be open to a better way of doing this if you can think of one.

**Comment by AydinHassan** [comment]
Actually that doesn't work, but this does:

```php
<?php

use PhpSchool\CliMenu\CliMenu;
use PhpSchool\CliMenu\Builder\CliMenuBuilder;
use PhpSchool\CliMenu\MenuItem\MenuMenuItem;

require_once(__DIR__ . '/../vendor/autoload.php');

$subMenu = function (CliMenuBuilder $b) {
    $b->setTitle('Behold the awesomeness')
        ->addItem('Print Hello', function () { echo 'Hello!'; });
};

$exit = function (CliMenu $menu) {
    $menu->close();
};

$menu = (new CliMenuBuilder)
    ->addSubMenu('[S]uper Sub Menu', $subMenu)
     ->build();

$menu->addCustomControlMapping('x', $exit);
$menu->addCustomControlMapping('s', function (CliMenu $menu) {
    foreach ($menu->getItems() as $item) {
        if ($item instanceof MenuMenuItem && $item->getText() === '[S]uper Sub Menu') {
            $item->showSubMenu($menu);
        }
    }
});

$menu->open();
```

**Comment by vesper8** [comment]
thanks a lot for finding a workaround! it's a bit wordy though.. guess I can wrap it in a method since I have many sub-menus I want to do this for. Maybe in a future version it could be built-in. If the 2nd argument passed to addCustomControlMapping is an instance of CliMenuBuilder then iterate through the menu's items and find the one with the matching title?

Another thing I wanted to ask about.. maybe I should create another issue for it.. is if we can have the ability to map keyboard shorcuts based on the current visible menu

Since I have a lot of sub-menus, I might want to re-use keyboard letters based on the menu that is currently shown to the user. I have to admit I haven't tried this yet so maybe it already works by default.

**Comment by AydinHassan** [comment]
Hey yeah I'd be happy to accept a PR to make that easier. Unfortunately It's not something I will be doing myself. Another option would be to just accept the shortcut key as a third parameter to `addSubMenu`. 

Regarding the second thing, I'm unsure, try it :)

**Comment by vesper8** [comment]
I might submit a PR later for the other things, for now I've hacked up this quick method that scans all menu items for letters in between square brackets and automatically assigns keyboard shortcuts to those it finds. I did very little testing but this demo works

```
        $subMenu1 = function (CliMenuBuilder $b) {
            $b->setTitle('Behold the awesomeness #1')
                ->addItem('Print Hello', function () { echo 'Hello!'; });
        };

        $subMenu2 = function (CliMenuBuilder $b) {
            $b->setTitle('Behold the awesomeness #2')
                ->addItem('Print Hello', function () { echo 'Hello!'; });
        };

        $menu = (new CliMenuBuilder)
        ->addSubMenu('[S]uper Sub Menu #1', $subMenu1)
        ->addSubMenu('S[u]per Sub Menu #2', $subMenu2)
        ->build();

        $this->autoMapKeyboardShortcuts($menu);

    public function autoMapKeyboardShortcuts($menu)
    {
        foreach ($menu->getItems() as $item) {
            if ($item instanceof \PhpSchool\CliMenu\MenuItem\MenuMenuItem && preg_match('/\[(.*?)\]/i', $item->getText(), $match)) {
                $menuItemText = $item->getText();
                $shortcutKey = strtolower($match[1]);

                $menu->addCustomControlMapping($shortcutKey, function (\PhpSchool\CliMenu\CliMenu $menu) use ($menuItemText) {
                    foreach ($menu->getItems() as $item) {
                        if ($item instanceof \PhpSchool\CliMenu\MenuItem\MenuMenuItem && $item->getText() === $menuItemText) {
                            $item->showSubMenu($menu);
                        }
                    }
                });
            }
        }
    }
```

Not the prettiest but useful.. let me know if you want a PR for that

**Comment by AydinHassan** [comment]
I like the idea! @Lynesth @mikeymike what do you think? I think some people might think it's a bit magic, but I dig it, so just checking with others.

**Comment by vesper8** [comment]
I started using it and noticed that I can't bind the letter 'm' ? it's somehow.. reserved?

```
        $subMenu1 = function (CliMenuBuilder $b) {
            $b->setTitle('Behold the awesomeness #1')
                ->addItem('Print Hello', function () { echo 'Hello!'; });
        };

        $subMenu2 = function (CliMenuBuilder $b) {
            $b->setTitle('Behold the awesomeness #2')
                ->addItem('Print Hello', function () { echo 'Hello!'; });
        };

        $menu = (new CliMenuBuilder)
        ->addSubMenu('[S]uper Sub Menu #1', $subMenu1)
        ->addSubMenu('Super Sub [M]enu #2', $subMenu2)
        ->build();

        $this->autoMapKeyboardShortcuts($menu);

        $menu->open();
```

```
InvalidArgumentException  : Cannot rebind this input
```

I can't find anywhere in your source where the letter 'm' is bound and pressing it does nothing.. so I'm confused.. works with other letters.. though I haven't tried the whole alphabet

**Comment by mikeymike** [comment]
definitely something of value imo, would need a little documentation to go with as I can imagine there's a few caveats with nested menus.

**Comment by mikeymike** [comment]
@vesper8 https://github.com/php-school/cli-menu/blob/master/src/CliMenu.php#L74

**Comment by vesper8** [comment]
ahh.. I see.. are those like.. VIM mappings? seems weird.. I would like to be able to disable that or override them

I also just improved my magic method to work with selectable items as well as menu items

```
        $subMenu1 = function (CliMenuBuilder $b) {
            $b->setTitle('Behold the awesomeness #1')
                ->addItem('Print Hello', function () { echo 'Hello!'; });
        };

        $subMenu2 = function (CliMenuBuilder $b) {
            $b->setTitle('Behold the awesomeness #2')
                ->addItem('Print Hello', function () { echo 'Hello!'; });
        };

        $menu = (new CliMenuBuilder)
        ->addSubMenu('[S]uper Sub Menu #1', $subMenu1)
        ->addSubMenu('S[u]per Sub Menu #2', $subMenu2)
        ->addItem('[A]', function () { echo 'A'; })
        ->build();

        $this->autoMapKeyboardShortcuts($menu);

        $menu->open();

    public function autoMapKeyboardShortcuts($menu)
    {
        foreach ($menu->getItems() as $item) {
            if (preg_match('/\[(.*?)\]/i', $item->getText(), $match)) {
                $menuItemText = $item->getText();
                $shortcutKey = strtolower($match[1]);

                $menu->addCustomControlMapping($shortcutKey, function (\PhpSchool\CliMenu\CliMenu $menu) use ($menuItemText) {
                    foreach ($menu->getItems() as $item) {
                        if ($item->getText() === $menuItemText) {
                            switch (true) {
                                case $item instanceof \PhpSchool\CliMenu\MenuItem\MenuMenuItem:
                                    $item->showSubMenu($menu);
                                break;

                                case $item instanceof \PhpSchool\CliMenu\MenuItem\SelectableItem:
                                    $callable = $item->getSelectAction();
                                    $callable($menu);
                                break;
                            }
                        }
                    }
                });
            }
        }
    }
```

**Comment by Lynesth** [comment]
@vesper8 
That's very unfortunate as I cannot find the code I wrote a while ago for a client using CliMenu since it was pretty much doing that.

Anyway, here is another way of doing it, defining each menu separatly from children to parents
```php
use PhpSchool\CliMenu\CliMenu;
use PhpSchool\CliMenu\Action\GoBackAction;
use PhpSchool\CliMenu\Builder\CliMenuBuilder;

require_once 'vendor/autoload.php';

$subMenuBuilder1 = (new CliMenuBuilder)
        ->disableDefaultItems();
        ->setTitle('CLI Menu > Options')
        ->addItem('First option', function () { echo ("I am an option"); })
        ->addLineBreak('-')
        ->addItem('Go Back', new GoBackAction); // Adding the "Go Back" manually, you can also add the "Exit" if wanted.
$subMenu1 = $subMenuBuilder1->build();

$mainMenuBuilder = (new CliMenuBuilder)
        ->setTitle('CLI Menu')
        ->addItem('First Item', function() { echo "Hello"; })
        ->addStaticItem('Second Item')
        ->addLineBreak('-')
        ->addSubMenuFromBuilder('Options', $subMenuBuilder1)
        ->setWidth(70);
$mainMenu = $mainMenuBuilder->build();

$mainMenu->addCustomControlMapping('o', function(CliMenu $menu) use ($subMenu1) {
        // Closing parent and opening sub
        $menu->closeThis();
        $subMenu1->open();
});

$mainMenu->open();
```

We could improve the whole thing I agree...

**Comment by AydinHassan** [comment]
Yeah I think a method to disable vim mappings should be fine. I would like to see the method `autoMapKeyboardShortcuts` on the CliMenuBuilder though if at all possible, and it would be nice to set it once on the parent builder and not have to set it again on each child builder.

**Comment by vesper8** [comment]
going to fork for now as I'm sure I'll find more bugs/improvements as I continue to play around with this package. Once things are a bit more stable I'll submit some PRs

**Comment by AydinHassan** [comment]
Sounds great @vesper8, looking forward to new features :)

**Comment by vesper8** [comment]
hrmm.. what do you mean they won't work? It's still calling it in this example:

```
        $subMenu = function (CliMenuBuilder $b) {
            $b->setTitle('Behold the awesomeness')
                ->addItem('Print Hello', function () { echo 'Hello!'; });
        };

        $itemCallable = function (CliMenu $menu) {
            echo $menu->getSelectedItem()->getText();
        };

        $menu = (new CliMenuBuilder)
        ->addSubMenu('[S]uper Sub Menu', $subMenu)
        ->addItem('[A]', function () { echo 'A'; })
        ->addItem('[B]', $itemCallable)
        ->build();

        $this->autoMapKeyboardShortcuts($menu);

        $menu->open();

    public function autoMapKeyboardShortcuts($menu)
    {
        foreach ($menu->getItems() as $item) {
            if (preg_match('/\[(.)\]/', $item->getText(), $match)) {
                $menu->addCustomControlMapping(strtolower($match[1]), function (\PhpSchool\CliMenu\CliMenu $menu) use ($item) {
                    if ($item->canSelect() && $callable = $item->getSelectAction()) {
                        $callable($menu);
                    }
                });
            }
        }
    }
```

**Comment by Lynesth** [comment]
@vesper8 I mean that if you are on the first element of the menu ("[S]uper Sub Menu" in your case) and then press `B`, it will echo "[S]uper Sub Menu", not the expected "[A]" as if you actually selected it and pressed `ENTER`.

**Comment by vesper8** [comment]
I see.. well I think if you have a callable that is suppose to interact with the currently highlighted menu item, then you just shouldn't add a global [key] shortcut to it because that doesn't make sense

**Comment by Lynesth** [comment]
@vesper8 Yes, I know. I was just pointing out that examples wouldn't work with this method :)

**Comment by Lynesth** [comment]
@vesper8 You'll also need to make it recursive in order for it to work on SplitItem too.
Here is a draft of the thing:

```php
function autoMapControls($menu, $items) {
    foreach ($items as $key => $item) {
        if ($item instanceof \PhpSchool\CliMenu\MenuItem\SplitItem) {
            autoMapControls($menu, $item->getItems());
            continue;
        }
        if (preg_match('/\[(.)\]/', $item->getText(), $match)) {
            $shortcutKey = strtolower($match[1]);

            $menu->addCustomControlMapping($shortcutKey, function (\PhpSchool\CliMenu\CliMenu $menu) use ($item) {
                if ($item->canSelect() && $callable = $item->getSelectAction()) {
                    $callable($menu);
                }
            });
        }
    }
}

autoMapControls($menu, $menu->getItems());
```

**Comment by vesper8** [comment]
@Lynesth thanks! yes I was just realizing that this doesn't work for submenus either.. which is more of concern

it only works if I call ->build() inside the submenu, but then that adds an extra set of 'go back' and 'exit'

**Comment by Lynesth** [comment]
@vesper8 There you go for submenus (so hopefully full recursiveness):

```php
function autoMapControls($menu, $items) {
    foreach ($items as $key => $item) {
        if ($item instanceof \PhpSchool\CliMenu\MenuItem\SplitItem) {
            autoMapControls($menu, $item->getItems());
            continue;
        }

        if (preg_match('/\[(.)\]/', $item->getText(), $match)) {
            $shortcutKey = strtolower($match[1]);

            $menu->addCustomControlMapping($shortcutKey, function (\PhpSchool\CliMenu\CliMenu $menu) use ($item) {
                if ($item->canSelect() && $callable = $item->getSelectAction()) {
                    $callable($menu);
                }
            });

            if ($item instanceof \PhpSchool\CliMenu\MenuItem\MenuMenuItem) {
                $submenu = $item->getSubMenu();
                autoMapControls($submenu, $submenu->getItems());
            }
        }
    }
}

autoMapControls($menu, $menu->getItems());
```

Now it's way too late here, I'm going to bed, have fun :)

**Comment by vesper8** [comment]
excellent!! this works very nicely!

```
        $subMenu = function (CliMenuBuilder $b) {
            $b->setTitle('Behold the awesomeness')
                ->addItem('[P]rint Hello', function () { echo 'Hello!'; });
        };

        $menu = (new CliMenuBuilder)
        ->addSubMenu('[S]uper Sub Menu #1', $subMenu)
        ->addItem('[A]', function () { echo 'A'; })
        ->addSplitItem(function (SplitItemBuilder $b) {
            $b->setGutter(5)
                    ->addSubMenu('Sub Menu [o]n a split item', function (CliMenuBuilder $b) {
                        $b->setTitle('Behold the awesomeness')
                            ->addItem('This is awesom[e]', function () { print 'Yes!'; });
                    })
                    ->addItem('Item [2]', function () { echo 'Item 2'; });
        })
        ->build();

        $this->autoMapKeyboardShortcuts($menu, $menu->getItems());

        $menu->open();

        return;

    public function autoMapKeyboardShortcuts($menu, $items)
    {
        foreach ($items as $item) {
            if ($item instanceof \PhpSchool\CliMenu\MenuItem\SplitItem) {
                $this->autoMapKeyboardShortcuts($menu, $item->getItems());
                continue;
            }

            if (preg_match('/\[(.)\]/', $item->getText(), $match)) {
                $shortcutKey = strtolower($match[1]);

                $menu->addCustomControlMapping($shortcutKey, function (\PhpSchool\CliMenu\CliMenu $menu) use ($item) {
                    if ($item->canSelect() && $callable = $item->getSelectAction()) {
                        $callable($menu);
                    }
                });

                if ($item instanceof \PhpSchool\CliMenu\MenuItem\MenuMenuItem) {
                    $submenu = $item->getSubMenu();
                    $this->autoMapKeyboardShortcuts($submenu, $submenu->getItems());
                }
            }
        }
    }
```

**Comment by vesper8** [comment]
@Lynesth I think it's almost perfect now.. only thing that's missing is allowing keyboard shortcuts to only apply to the current menu, so keys could be re-used.

I think your idea about building menus separately and opening/closing them instead of using the concept of sub-menus could allow this to be possible.. maybe that is the best way.. this could make it difficult to manage though when going back to parent menu.. but it could all tie in nicely if structured correctly

maybe there's a better way though, one that works with the latest example posted above. We could change the constraints to allow multiple items to be mapped to the same key, but somehow track which menu we are on when hitting a key and only execute the one that matches the current menu.

Would like to hear your thoughts on this seeing as you've been so quick with putting out great code to meet these new challenges : )

**Comment by Lynesth** [comment]
@vesper8 What do you mean ?
Shortcuts are menu based which mean that you can have the same key used for different menus. Does that not work for you ?

**Comment by vesper8** [comment]
oh wow! damn.. my bad totally should have tested that before.. it does work!! awesomely well in fact!

**Comment by vesper8** [comment]
one more for you :)

I want to map a key to the 'go back' on each submenu. I've managed this by adding this line to the autoMap method:

```
$submenu->addCustomControlMapping('b', new \PhpSchool\CliMenu\Action\GoBackAction);
```

```
    public function autoMapKeyboardShortcuts($menu, $items)
    {
        foreach ($items as $item) {
            if ($item instanceof \PhpSchool\CliMenu\MenuItem\SplitItem) {
                $this->autoMapKeyboardShortcuts($menu, $item->getItems());
                continue;
            }

            if (preg_match('/\[(.)\]/', $item->getText(), $match)) {
                $shortcutKey = strtolower($match[1]);

                $menu->addCustomControlMapping($shortcutKey, function (\PhpSchool\CliMenu\CliMenu $menu) use ($item) {
                    if ($item->canSelect() && $callable = $item->getSelectAction()) {
                        $callable($menu);
                    }
                });

                if ($item instanceof \PhpSchool\CliMenu\MenuItem\MenuMenuItem) {
                    $submenu = $item->getSubMenu();
                    $this->autoMapKeyboardShortcuts($submenu, $submenu->getItems());
                    $submenu->addCustomControlMapping('b', new \PhpSchool\CliMenu\Action\GoBackAction);
                }
            }
        }
    }
```

it work nicely, but I would like to map it to "meta - up arrow".. that is.. on the mac.. CMD-UP

Do you know how/if I can map it to that?

**Comment by vesper8** [comment]
By printing out the $rawChar inside the CliMenu.php I managed to make it work with "ALT-UP" with this mapping:

```
$submenu->addCustomControlMapping("\x1b\x1b[A", new \PhpSchool\CliMenu\Action\GoBackAction);
```

I don't think I can make it work with CMD-UP since that's a registered hotkey inside iTerm2

I also wonder if there's a way to break out of "askText()", ESC doesn't work.. CTRL-C seems to be the only way.. or answering the question but there should be a way to cancel ideally

**Comment by Lynesth** [comment]
@vesper8 
I really don't think we can do anything with `CMD + Key` or `WIN + Key` since they are not printed out.
For the `ESC` part, you may have a look at #174

**Comment by Lynesth** [comment]
For the Go Back shortcut, wouldn't it work to simply change the `goBackButtonText` to something like "Go [B]ack" ?

**Comment by vesper8** [comment]
yes that would be one way to handle it, however I like navigating with CMD-UP/CMD-DOWN rather than reaching for the enter and b keys

For the "go back" behaviour I managed it, As I mentioned above, by adding (now mapped to ALT-UP)

```
$submenu->addCustomControlMapping("\x1b\x1b[A", new \PhpSchool\CliMenu\Action\GoBackAction);
```

For mapping ALT-DOWN to "Enter" I added

```
$menu->setDefaultControlMappings([
    "\x1b\x1b[B" => \PhpSchool\Terminal\InputCharacter::ENTER,
]);
```

But then I noticed that doing so only makes it work on the parent menu.. so I also added it to the submenu in the automap method

```
$submenu->setDefaultControlMappings([
    "\x1b\x1b[B" => \PhpSchool\Terminal\InputCharacter::ENTER,
]);
```

So I guess that user-defined defaults only apply on a per-menu basis as opposed to the defaults that were hard-coded.. not ideal

Superb about the ESC key btw!!

**Comment by AydinHassan** [comment]
@vesper8 can this be closed now we have #176 ?

**Comment by vesper8** [comment]
yes of course! really glad this prompted all this work to happen! great team effort!
